### PR TITLE
Group-Object - Use Case-Sensitive Hashtable for -CaseSensitive -AsHashtable

### LINKED ISSUE #10441: Return casesensitive hashtable with Group-Object
# Steps to reproduce
I know PowerShell is mostly case insensitive, but when using both `-AsHashTable` and `-CaseSensitive` I expected to end up with a case sensitive hashtable.
```powershell
$capitonyms = @(
    [PSCustomObject]@{
        Capitonym = 'Bill'
    }
    [PSCustomObject]@{
        Capitonym = 'bill'
    }
)

$capitonyms | Group-Object Capitonym -AsHashTable -CaseSensitive
```

# Expected behavior

```none
Name                           Value
----                           -----
Bill                           {@{Capitonym=Bill}}
bill                           {@{Capitonym=bill}}
```

# Actual behavior

```none
Group-Object : The objects grouped by this property cannot be expanded because there is a key duplication. Provide a valid value for the property, and then try again.
At line:1 char:15
+ $capitonyms | Group-Object Capitonym -AsHashTable -CaseSensitive
+               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : InvalidArgument: (:) [Group-Object], Exception
+ FullyQualifiedErrorId : The objects grouped by this property cannot be expanded because there is a key duplication. Provide a valid value for the property, and then try again.,Microsoft.PowerShell.Commands.GroupObjectCommand
```

# Environment data

```none
Name                           Value
----                           -----
PSVersion                      7.0.0-preview.3
PSEdition                      Core
GitCommitId                    7.0.0-preview.3
OS                             Microsoft Windows 10.0.18362
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### PR DESCRIPTION
**Pull request by vexx32**
# PR Summary

Prior to this change, `Group-Object -AsHashtable -CaseSensitive` would give a key duplication error when given entries that only differ by casing. This was due to always using a case-insensitive hashtable, despite the request for `-CaseSensitive` behaviour.

This change allows `Group-Object -CaseSensitive -AsHashtable` to create and return a case-sensitive hashtable to enable this use case.

## PR Context

Resolves #10441

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/Group-Object.cs [hunk 1]
```diff
@@ -518,7 +518,16 @@ protected override void EndProcessing()
...
                 {
-                    Hashtable hashtable = CollectionsUtil.CreateCaseInsensitiveHashtable();
+                    Hashtable hashtable;
+                    if (CaseSensitive.IsPresent)
+                    {
+                        hashtable = new Hashtable();
+                    }
+                    else
+                    {
+                        hashtable = CollectionsUtil.CreateCaseInsensitiveHashtable();
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/Group-Object.cs]: I think we can simplify the code and this helps us to add Culture parameter later #9348:
```c#
var comparer = CaseSensitive.IsPresent ? StringComparer.CurrentCulture : StringComparer.CurrentCultureIgnoreCase;
hashtable = Hashtable(comparer);

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Group-Object.Tests.ps1 [hunk 2]
```diff
@@ -7,9 +7,9 @@ Describe "Group-Object DRT Unit Tests" -Tags "CI" {
         $results.Count | Should -Be 2
         $results.Name.Count | Should -Be 2
         $results.Group.Count | Should -Be 4
-        $results.Name | Should -Be aA,AA
-        $results.Group | Should -Be aA,aA,AA,AA
-        ,$results | Should -BeOfType "System.Array"
+        $results.Name | Should -Be aA, AA
+        $results.Group | Should -Be aA, aA, AA, AA
+        , $results | Should -BeOfType "System.Array"
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Group-Object.Tests.ps1]: Please move style fixes in another PR. Below too.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Group-Object.Tests.ps1 [hunk 3]
```diff
@@ -173,6 +173,23 @@ Describe "Check 'Culture' parameter in order object cmdlets (Group-Object, Sort-
...
+                }
+                [PSCustomObject]@{
+                    Capitonym = 'bill'
+                }
+            )
+
+            $capitonyms | Group-Object Capitonym -AsHashTable -CaseSensitive
+        }
+
+        $script | Should -Not -Throw
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Group-Object.Tests.ps1]: I'd prefer positive checks.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Group-Object.Tests.ps1 [hunk 4]
```diff
@@ -173,6 +173,23 @@ Describe "Check 'Culture' parameter in order object cmdlets (Group-Object, Sort-
...
+            $capitonyms = @(
+                [PSCustomObject]@{
+                    Capitonym = 'Bill'
+                }
+                [PSCustomObject]@{
+                    Capitonym = 'bill'
+                }
+            )
+
+            $capitonyms | Group-Object Capitonym -AsHashTable -CaseSensitive
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Group-Object.Tests.ps1]:
```suggestion
            $capitonyms | Group-Object -Property Capitonym -AsHashTable -CaseSensitive
```

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/Group-Object.cs [hunk 5]
```diff
@@ -518,7 +518,16 @@ protected override void EndProcessing()
             {
                 if (AsHashTable)
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/Group-Object.cs]: For consistency:
```suggestion
                 if (AsHashTable.IsPresent)
```

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/Group-Object.cs (hunk 1 above)
Comment (vexx32) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/Group-Object.cs]: Can do! I'm not sure if these methods use currentculture or ordinal though. I'll check :)

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Group-Object.Tests.ps1 (hunk 2 above)
Comment (vexx32) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/Group-Object.Tests.ps1]: VS code's autoformat strikes again. It is in a separate commit at least, but if you'd prefer to do a whole separate PR I can manage. :slightly_smiling_face:

**Comment by vexx32** [comment]
@iSazonov I think that should address all your concerns. I'll look at doing a style pass over the tests and possible the cmdlet after this PR is completed. 😊

**Review by iSazonov — APPROVED** [review]

**Comment by iSazonov** [comment]
@vexx32 I don't know should we document this in docs repo?

**Comment by vexx32** [comment]
🤔 I suppose it _should_ be documented since we're changing established behaviour. I'll open an issue.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Group-Object.Tests.ps1 [hunk 6]
```diff
@@ -173,6 +173,20 @@ Describe "Check 'Culture' parameter in order object cmdlets (Group-Object, Sort-
...
+        $capitonyms = @(
+            [PSCustomObject]@{
+                Capitonym = 'Bill'
+            }
+            [PSCustomObject]@{
+                Capitonym = 'bill'
+            }
+        )
+
+        [hashtable] $Result = $capitonyms | Group-Object -Property Capitonym -AsHashTable -CaseSensitive
```
Comment (SteveL-MSFT) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Group-Object.Tests.ps1]: If this gets regressed in the future, I think this will give a better test error message

```suggestion
        $Result = $capitonyms | Group-Object -Property Capitonym -AsHashTable -CaseSensitive
        $Result | Should -BeOfType [HashTable]
```

**Review by SteveL-MSFT — CHANGES_REQUESTED** [review]

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Group-Object.Tests.ps1 (hunk 6 above)
Comment (vexx32) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/Group-Object.Tests.ps1]: Good catch! 🙂

**Review by SteveL-MSFT — APPROVED** [review]

**Comment by vexx32** [comment]
@SteveL-MSFT 🤔 do you know if the bot ignores non-required checks? Pretty sure it's waiting on the CodeFactor issue (which is just "complex method" on EndProcessing() here.

I'm happy to refactor it out into helper methods if you feel that's necessary. 🙂
